### PR TITLE
first implementation of faulty zero filter fz_filter

### LINKED ISSUE #2: Implement FZ (faulty zero) filter from PWSQC
A Python implementation of the FZ (faulty zero) filter was provided for the OPENSENSE training school [here](https://github.com/OpenSenseAction/training_school_opensene_2023/blob/b87ea029317f70155021fbcd5e90424932031d5a/notebooks_pws/FZ_filter.py).

As first step this function should be added here, but with a reasonable docstring. A test should be added. If this works, we can later, in a separate issue and separate PR, potentially work on speeding up the implementation by avoiding the for-loop.

### PR DESCRIPTION
**Pull request by cchwala**
This will close #2 

TODOs from WIP state from last week (copy-paste from comment below):

some TODOs
- [x] add docstring
- [x] understand why the `np.ma.compressed` code is there and potentially remove it if it is not needed (which would be my guess now)
- [x] add a second test with some variation (not sure what edge cases could be there) <-- added one additional test, but there might remain edge cases to be checked, which, however, requires more discussion about the function and hence is postponed for a future PR
- [x] add correct type hints (not sure if `npt.NDArray` is good enough) <-- marked as done, but should be checked again later if this is correct
- [x] add a simple example notebook with the usage of the filter
- [x] rename module file (and decide in general how module file structure should look like)

### CODE REVIEW

**Comment by cchwala** [comment]
Since test are not running, because linting already fails, I put a quick note about test coverage here before I change parts of the code. Only L23 and L25 are not covered by tests.

https://github.com/OpenSenseAction/pypwsqc/blob/110c790ccc95fdc5baedc244d79dd62ef6ddaf1a/src/pypwsqc/fz_filter.py#L22-L25

I am not sure what `np.ma.compressed ... = 0` things should actually do. It looks a bit like checking for all NaNs.

**Comment by cchwala** [comment]
some TODOs
- [ ] add docstring
- [ ] understand why the `np.ma.compressed` code is there and potentially remove it if it is not needed (which would be my guess now)
- [ ] add a second test with some variation (not sure what edge cases could be there)
- [ ] add a simple example notebook with the usage of the filter
- [ ] rename module file (and decide in general how module file structure should look like)

**Comment by LottedeVos** [comment]
> Since test are not running (UPDATE: test now run, see below, and codecov shows that these lines are note covered), because linting already fails, I put a quick note about test coverage here before I change parts of the code. Only L23 and L25 are not covered by tests.
> 
> https://github.com/OpenSenseAction/pypwsqc/blob/110c790ccc95fdc5baedc244d79dd62ef6ddaf1a/src/pypwsqc/fz_filter.py#L22-L25
> 
> I am not sure what the purpose of the `np.ma.compressed ... = 0` things is. I do not see masked arrays being used above this code. Maybe there where some masked arrays in the code where the FZ_filter function originally came from.

Honestly I'm not quite sure, I think Jochen will be able to provide more detail as it is his translation of the code. Based on what I could quickly google: I suspect is has to do with that FZ_array is first defined as an array of '-1' values. If the Ref_array does not contain a value, the FZ_array should remain -1.  If Ref_array does contain a value but the Sensor_array does, the FZ_array receives the value that it had on the previous interval. 

Basically: if you don't have a reference, you cannot attribute a FZ-flag (i.e. FZ flag is -1). If you don't have a value in your PWS time series, the FZ flag of the previous time interval is attributed. 

I hope this helps!

**Comment by cchwala** [comment]
Thanks. That helps to understand the logic.

But the part with `np.ma.compressed` is very specific because it only does something useful if e.g. `Ref_array` is a ["masked" numpy array](https://numpy.org/doc/stable/reference/maskedarray.html).

I did some test to check the behavior of `np.ma.compressed`:

Bildschirmfoto 2024-01-24 um 11 04 41

This might stem from the usage of masked arrays, e.g. from radar data, that have been used with the Python code that we use here. Based on my current understanding of the code, I would just remove the two statements with `np.ma.compressed`.

@JochenSeidel Do you have any additional info?
